### ISSUE TITLE
v2 cannot find forked branches

### ISSUE THREAD

**Issue by dereuromark** [composer/composer#9482]
My `composer.json`:

```
    "require": {
        "burzum/cakephp-file-storage": "dev-metadata",
    ...
    "repositories": [
            {
                "type": "git",
                "url": "git@github.com:dereuromark/cakephp-file-storage.git"
            }
        ]
    "minimum-stability": "dev",
    "prefer-stable": true,
```
as fork from the original https://github.com/burzum/cakephp-file-storage

But composer v2.0.7:

> - Root composer.json requires burzum/cakephp-file-storage dev-metadata as 4.0, found burzum/cakephp-file-storage[dev-master, dev-0.0-cake2, 0.3.2, ..., 0.5.2, 1.0.0-RC1, ..., 1.2.x-dev, 2.0.0-alpha1, ..., 2.1, 3.0.0-beta, 3.0.x-dev, 4.0.x-dev] but it does not match the constraint.

It cannot find those explicit branches to test a forked branch.

I also tried inline aliasing: `"burzum/cakephp-file-storage": "dev-metadata as 4.0"` etc.

Is this fixed with https://github.com/composer/composer/pull/9477 or is that another issue?

**Comment by stof** [comment]
>     * Root composer.json requires burzum/cakephp-file-storage dev-metadata as 4.0,

This error message does not match the composer.json you shared.

**Comment by dereuromark** [comment]
I tried both alias and without alias. Same affect.

**Comment by Seldaek** [comment]
I can not repro with simply this file:

```
{
    "require": {
        "burzum/cakephp-file-storage": "dev-metadata"
    },
    "repositories": [
        {
            "type": "git",
            "url": "git@github.com:dereuromark/cakephp-file-storage.git"
        }
    ]
}
```

If you can repro with that, then please run a `composer update -vvv` and share the full output. Maybe it'll help.

**Comment by dereuromark** [comment]
I can definitly reproduce still with this file
```
{
    "require": {
        "php": ">=7.4.10",
        "ext-json": "*",
        "burzum/cakephp-file-storage": "dev-metadata",
        "burzum/cakephp-service-layer": "^2.0.0",
        "cakephp/cakephp": "~4.1.0",
        "cakephp/migrations": "^3.0",
        "cakephp/plugin-installer": "^1.2",
        "friendsofcake/bootstrap-ui": "^3.0",
        "friendsofcake/search": "dev-master as 6.0.0",
        "geocoder-php/google-maps-provider": "^4.4.0",
        "geocoder-php/provider-implementation": "^1.0",
        "geoip2/geoip2": "^2.0",
        "league/container": "^3.3",
        "linkorb/jsmin-php": "dev-master",
        "markstory/asset_compress": "@stable",
        "mobiledetect/mobiledetectlib": "^2.8",
        "natxet/cssmin": "dev-master",
        "phauthentic/file-storage-image-processor": "dev-develop as 1.0.0",
        "php-http/cakephp-adapter": "0.3.x-dev",
        "php-http/message": "^1.8.0"
    },
    "require-dev": {
        "cakephp/bake": "dev-master as 2.1.2",
        "dereuromark/cakephp-ide-helper": "dev-master as 1.8.0",
        "dereuromark/cakephp-ide-helper-extra": "dev-master",
        "spryker/code-sniffer": "@stable",
        "cakephp/debug_kit": "dev-master as 4.3.5",
        "josegonzalez/dotenv": "^3.2",
        "phpunit/phpunit": "~8.5.0"
    },
    "autoload": {
        "psr-4": {
            "App\\": "src/",
            "App\\Test\\Fixture\\": "tests/Fixture/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "App\\Test\\": "tests/",
            "Cake\\Test\\": "vendor/cakephp/cakephp/tests/"
        }
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "config": {
        "sort-packages": true,
        "platform": {
            "php": "7.4.10"
        },
        "process-timeout": 0
    }
}
```
Since I also tried in an empty fresh folder, it is not about lock file or existing vendor folder either.

**Comment by stof** [comment]
@dereuromark your reproducing case does not have the `repositories` configuration. so not finding the branch of your fork is totally expected.

**Comment by dereuromark** [comment]
I did and do.
I just tried to take out all non-relevant info for publishing and killed that part. I fixed it up in the above code snippet and double checked.
It definitely is not working and that is not expected :)

**Comment by Seldaek** [comment]
@dereuromark can you please try with my json?

I get output like this:

```
Reading composer.json of burzum/cakephp-file-storage (metadata)
Reading CACHE_DIR/repo/git-github.com-dereuromark-cakephp-file-storage.git/4f2f186504e0b72c2ab332cd7ac468fcbfa251ec from cache
Importing branch metadata (dev-metadata)
Reading composer.json of burzum/cakephp-file-storage (showcase-antipattern)
Reading CACHE_DIR/repo/git-github.com-dereuromark-cakephp-file-storage.git/f67de905cbf8d376af3a1f59e8a6651a30cfc631 from cache
Importing branch showcase-antipattern (dev-showcase-antipattern)
Downloading https://repo.packagist.org/packages.json
[200] https://repo.packagist.org/packages.json
```

Where you see it'si mporting the metadata branch from the custom repo, and finally it fails due to a related develop branch but that's fairly irrelevant:

```
  Problem 1
    - Root composer.json requires burzum/cakephp-file-storage dev-metadata -> satisfiable by burzum/cakephp-file-storage[dev-metadata].
    - burzum/cakephp-file-storage dev-metadata requires phauthentic/file-storage dev-develop -> found phauthentic/file-storage[dev-develop] but it does not match your minimum-stability.
```

**Comment by Seldaek** [comment]
Also this for shortened output:

```
$ composer update -vvv 2>&1 | grep Importing | grep metadata
Importing branch 4.0-metadata (dev-4.0-metadata)
Importing branch metadata (dev-metadata)
```

I feel like there's a user error or you are cleaning up too much when sharing info, but I definitely can not reproduce.

**Comment by dereuromark** [comment]
I can also reproduce even with your simple example above
Same issue, and the grep command displays 0 output, so there must be something wrong with composer.

I run 2.0.7 and I did also composer clearcache to make sure it ain't any cache issue.

**Comment by Seldaek** [comment]
Ok then it gets trickier.. Still interested in a complete -vvv output with my json, which should not contain any sensitive info. It would show what is going on when it tries to load the info from the git repo hopefully.

As a workaround, can you try with `type: vcs` instead of `type: git`? Then it should use github API to gather info instead of a clone.. That might fix it for you until we figure out what is going on.

**Comment by dereuromark** [comment]
Your example was missing:

    "minimum-stability": "dev",
    "prefer-stable": true

OK, when I add the missing config in your repo I get the output
```
Importing branch 4.0-metadata (dev-4.0-metadata)
Importing branch metadata (dev-metadata)
```

And then it works fine...
Really weird.

**Comment by dereuromark** [comment]
I dont have time to investigate further now and just use a normal branch on the same repo for simplicity.
I will then close this until we can more clearly state the reproducible issue.

**Comment by Seldaek** [comment]
But these minimum-stability/prefer-stable lines you added changes nothing to the repository scanning, it only gets the solver error removed. Anyway glad it works now but I still don't understand anything of what happened.

**Comment by tomkiss** [comment]
FWIW on this old thread, I couldn't fathom why my custom fork wasn't loading.

I looked at the output using `-vvv` and saw that the branch name was listed as `branchname-dev`. I then changed my require from `dev-branchname` to `branchname-dev`, and it worked.

Absolutely no idea why, but it fixed it so I'm posting here incase anyone else discovers this issue, using composer 2.4.1, php 8.1 on a ddev docker instance.
